### PR TITLE
Generic telemetry logs [APPSEC-10883]

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?
It moves the existing telemetry logs module inside the IAST to the tracer telemetry namespace but with the difference that it now expects the logs to come to it through `datadog:telemetry:log` channel.

The log objects sent via `datadog:telemetry:log` channel must meet the [telemetry log schema](https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/SchemaDocumentation/Schemas/log_message.md)

The `telemetry/logs` module only accepts  a log with one of the following levels: ERROR, WARN and DEBUG as it is specified in the [telemetry logs API spec](https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/producing-telemetry.md#logs). 
When log collection is enabled the default level is WARN and in the future it could be possible to enable DEBUG level but it is not covered yet.

After `telemetry/logs` module receives a log, it delegates on `telemetry/logs/log-collector` to calculate the log hash with the different log properties: `message`, `level` and `stack_trace` (`tags` and `tracer_time` are not used yet) and store it in a Map to avoid log duplication in the same batch of logs. 
After the log-collector is drained it is possible to store again the same log. 

NOTE: At the moment the `stack_trace` redaction is done by the sender of the log ([iast-log.js](https://github.com/DataDog/dd-trace-js/pull/3647/files#diff-52563b1a19f60530c80eead4c62c33be1ad0e49ff6623d04f011a53c82c2715fL11)) but it could be `telemetry/logs` responsibility.

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/config.js [hunk 1]
```diff
@@ -583,8 +583,8 @@ ken|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)
     this.telemetry = {
       enabled: DD_TRACE_EXPORTER !== 'datadog' && isTrue(DD_TRACE_TELEMETRY_ENABLED),
       heartbeatInterval: DD_TELEMETRY_HEARTBEAT_INTERVAL,
-      logCollection: isTrue(DD_TELEMETRY_LOG_COLLECTION_ENABLED),
       debug: isTrue(DD_TELEMETRY_DEBUG),
+      logCollection: isTrue(DD_TELEMETRY_LOG_COLLECTION_ENABLED),
```
Comment (uurien) [inline packages/dd-trace/src/config.js:587]: why this movement?

File: packages/dd-trace/src/telemetry/log/index.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+const dc = require('../../../../diagnostics_channel')
+const logCollector = require('./log-collector')
+const { sendData } = require('../send-data')
+
+const telemetryLog = dc.channel('datadog:telemetry:log')
+
+let enabled = false
+
+// TODO: set value when the env var is defined
+const debug = false
```
Comment (uurien) [inline packages/dd-trace/src/telemetry/log/index.js:12]: Could we remove the code related with this property until its definition?

File: packages/dd-trace/src/config.js (hunk 1 above)
Comment (iunanua) [reply packages/dd-trace/src/config.js:587]: To put together properties of telemetry components.

File: packages/dd-trace/src/telemetry/index.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ const dependencies = require('./dependencies')
 const { sendData } = require('./send-data')
 
 const { manager: metricsManager } = require('./metrics')
+const logs = require('./log')
```
Comment (uurien) [inline packages/dd-trace/src/telemetry/index.js:10]: requiring `log` to varname `logs`, sounds weird.

Comment (iunanua) [reply packages/dd-trace/src/telemetry/index.js:10]: I had to rename the module as `log` because `logs` are ignored in .gitignore but I'm about to rename it again as `logs` and include `telemetry/logs` in .gitignore. wdyt?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/index.js]: renamed :open_mouth:

File: packages/dd-trace/src/telemetry/logs/log-collector.js [hunk 4]
```diff
@@ -18,37 +18,24 @@ function hashCode (hashSource) {
...
 function createHash (logEntry) {
-  if (!logEntry) return 0
-
   const prime = 31
   let result = ((!logEntry.level) ? 0 : hashCode(logEntry.level))
   result = (((prime * result) | 0) + ((!logEntry.message) ? 0 : hashCode(logEntry.message))) | 0
 
-  // NOTE: tags are not used at the moment
+  // NOTE: tags not used yet but must be of type string and comma separated value in the form 'TAG:value, TAG:value'
   // result = (((prime * result) | 0) + ((!logEntry.tags) ? 0 : hashCode(logEntry.tags))) | 0
```
Comment (uurien) [inline packages/dd-trace/src/telemetry/logs/log-collector.js:26]: What is the goal of this lines? If we are not using tags yet, why do we need code commented and the NOTE?

Comment (iunanua) [reply packages/dd-trace/src/telemetry/logs/log-collector.js]: removed

**Review by uurien — APPROVED** [review]

**Review by uurien — APPROVED** [review]

File: .gitignore [hunk 5]
```diff
@@ -124,3 +124,4 @@ packages/datadog-plugin-next/test/yarn.lock
 packages/dd-trace/test/appsec/next/*/package.json
 packages/dd-trace/test/appsec/next/*/node_modules
 packages/dd-trace/test/appsec/next/*/yarn.lock
+!packages/dd-trace/**/telemetry/logs
```
Comment (rochdev) [inline .gitignore:127]: Why is this needed?

**Review by rochdev — APPROVED** [review]
Overall approach LGTM, although I would prefer to see non-telemetry-specific events that are being listened to by telemetry instead of explicitly using a telemetry channel. Basically, the different components of the library would just publish what they're doing on separate channels, and any subscriber can listen to that and do something with them, for example sending telemetry logs but that's just one example. The new approach is close enough to that however so we can move forward with it and readjust if we want later.

File: .gitignore (hunk 5 above)
Comment (iunanua) [reply .gitignore:127]: because `logs` is globally ignored a few lines above
